pkg/controllers/workspace_controller.go

```diff
@@ -55,94 +58,132 @@ func (c *WorkspaceReconciler) Reconcile(ctx context.Context, req reconcile.Reque
 
 func (c *WorkspaceReconciler) applyWorkspaceResource(ctx context.Context, wObj *kdmv1alpha1.Workspace) error {
 	klog.InfoS("applyWorkspaceResource", "workspace", klog.KObj(wObj))
-	// Check CandidateNodes, if the count, instance type
-	candidateNodeCount, err := c.validateCandidateNodes(ctx, wObj)
+
+	var validNodeList []*corev1.Node
+
+	// Check the current cluster nodes if they match the labelSelector and instanceType
+	validCurrentClusterNodeList, err := c.validateCurrentClusterNodes(ctx, wObj)
 	if err != nil {
 		return err
 	}
-
-	remainingNodeCount := lo.FromPtr(wObj.Resource.Count) - candidateNodeCount
-	// if current node Count == workspace count, then all good and return
-	if remainingNodeCount <= 0 {
-		klog.InfoS("number of candidate nodes, %d, is equal or greater than required workspace count, %d ",
-			candidateNodeCount, lo.FromPtr(wObj.Resource.Count))
-		return nil
+	validNodeList = append(validNodeList, validCurrentClusterNodeList...)
+
+	// Check PreferredNodes
+	preferredList := wObj.Resource.PreferredNodes
+	if preferredList == nil {
+		klog.InfoS("PreferredNodes list is empty")
+	} else {
+		for n := range preferredList {
+			lo.Find(validNodeList, func(item *corev1.Node) bool {
+				if item.Name == preferredList[n] {
+					return true
+				}
+				if len(validNodeList) > lo.FromPtr(wObj.Resource.Count) {
+					validNodeList = append(validNodeList[:n], validNodeList[n+1:]...)
```

Comment:
Can you start from an empty validNodeList and add when a preferred node in the validCurrentClusterNodeList,?